Update existing credentials routes

valya opened the pull request.

BHesseldieck committed:
🎉 initial design

BHesseldieck committed:
✨ sharing/unsharing of credentials

Commit:
:recycle: split credential update route into controller and service

Commit:
:fire: remove credentials test that is no longer applicable

Commit:
:recycle: split credential creation route into controller and service

Commit:
:recycle: split single credential get

Commit:
:recycle: split delete credentials route

Commit:
:recycle: split get all credentials route

Commit:
:fire: remove unused imports in credentials contoller

Commit:
Merge remote-tracking branch 'origin/user-management-p2' into n8n-3959-update-existing-credentials-routes-12

ivov commented on packages/cli/config/schema.ts:
```diff
@@ -852,6 +852,12 @@ export const schema = {
 			default: 'default',
 			env: 'N8N_DEPLOYMENT_TYPE',
 		},
+		paid: {
+			doc: 'Whether paid features are enabled.',
+			format: Boolean,
+			default: true,
+			env: 'N8N_PAID',
+		},
```
Maybe a comment that this is a temp placeholder?

ivov commented on packages/cli/src/credentials/credentials.controller.ee.ts:
```diff
@@ -0,0 +1,62 @@
+/* eslint-disable import/no-cycle */
```
Review for cred services and controllers here: https://github.com/n8n-io/n8n/pull/3601

ivov commented on packages/cli/src/credentials/credentials.controller.ts:
```diff
@@ -0,0 +1,245 @@
+		]);
+
+		if (!shared) {
+			console.error('cannae update');
```
```suggestion

```

ivov commented on packages/cli/src/credentials/credentials.service.ts:
```diff
@@ -0,0 +1,259 @@
+	}
+
+	static async prepareCredentialsUpdateData(
+		data: CredentialRequest.RequestBody,
```
`CredentialRequest.CredentialProperties`?

ivov commented on packages/cli/src/credentials/credentials.service.ts:
```diff
@@ -0,0 +1,259 @@
+	}
+
+	static async prepareCredentialsCreateData(
+		data: CredentialRequest.RequestBody,
```
`CredentialRequest.CredentialProperties`?

ivov commented on packages/cli/src/user/user.service.ts:
```diff
@@ -0,0 +1,8 @@
+import { Db } from '..';
+import { User } from '../databases/entities/User';
```
```suggestion
import type { User } from '../databases/entities/User';
```

ivov commented on packages/cli/src/user/user.service.ts:
```diff
@@ -0,0 +1,8 @@
+	static async get(user: Partial<User>): Promise<User | undefined> {
+		return Db.collections.User.findOne(user);
+	}
+}
```
Both imports seem to be triggering `import/no-cycle`.

ivov commented on packages/cli/src/role/role.service.ts:
```diff
@@ -0,0 +1,8 @@
+import { Db } from '..';
+import { Role } from '../databases/entities/Role';
```
```suggestion
import type { Role } from '../databases/entities/Role';
```

ivov commented on packages/cli/src/role/role.service.ts:
```diff
@@ -0,0 +1,8 @@
+	static async get(role: Partial<Role>): Promise<Role | undefined> {
+		return Db.collections.Role.findOne(role);
+	}
+}
```
Both imports seem to be triggering `import/no-cycle`.

ivov commented on packages/cli/test/integration/credentials.api.test.ts:
```diff
@@ -2,6 +2,7 @@ import express from 'express';
 import { UserSettings } from 'n8n-core';
 import { Db } from '../../src';
 import { randomName, randomString } from './shared/random';
+import config from '../../config';
```
Separate PR https://github.com/n8n-io/n8n/pull/3601 introduces `authAgent`, can also be used to reduce boilerplate here.

ivov commented on packages/cli/test/integration/shared/testDb.ts:
```diff
@@ -681,7 +681,7 @@ export const getMySqlOptions = ({ name }: { name: string }): ConnectionOptions =
 	const encryptionKey = await UserSettings.getEncryptionKey();
 
-	const coreCredential = createCredentiasFromCredentialsEntity(credential, true);
+	const coreCredential = createCredentialsFromCredentialsEntity(credential, true);
```
If what we get is a core credential, then maybe `createCoreCredentialFromEntity()` or just `createCoreCredential`?

ivov reviewed (commented):
Great work! All tests passing on all three DBs. All inline comments only nitpicks.

General comments:

1. Top level inside `/src` now has a mixture of service-controllers and other dirs with full features like UM. Should we keep these service-controllers dirs under a single dir, as we used to have with `/api`? As opposed to scattered at top level:

```
/src
  /credentials
  /user
  /role
  /PublicApi
  /UserManagement
```

2. All test suites still have an `api` infix but we no longer have an `/api` dir. Perhaps naming should be updated, if we decide on a single dir for services-controllers?

```
packages/cli/test/integration/credentials.api.test.ts
```

3. Getting this on build:

```
src/credentials/credentials.controller.ee.ts(29,40): error TS2345: Argument of type 'User | undefined' is not assignable to parameter of type 'User'.
  Type 'undefined' is not assignable to type 'User'.
src/credentials/credentials.controller.ee.ts(53,55): error TS2345: Argument of type 'User | undefined' is not assignable to parameter of type 'User'.
  Type 'undefined' is not assignable to type 'User'.
src/credentials/credentials.service.ee.ts(14,61): error TS2345: Argument of type 'string' is not assignable to parameter of type 'User'.
src/credentials/credentials.service.ts(100,27): error TS2694: Namespace '"/Users/ivov/Desktop/n8n2/packages/cli/src/requests".CredentialRequest' has no exported member 'RequestBody'.
src/credentials/credentials.service.ts(120,27): error TS2694: Namespace '"/Users/ivov/Desktop/n8n2/packages/cli/src/requests".CredentialRequest' has no exported member 'RequestBody'.
```

Commit:
:fire: remove console.log

Commit:
Merge remote-tracking branch 'origin/user-management-p2' into n8n-3959-update-existing-credentials-routes-12

ivov commented on packages/cli/src/credentials/credentials.service.ee.ts:
```diff
@@ -8,10 +8,10 @@ import { RoleService } from '../role/role.service';
 		credentialId: string,
 	): Promise<{ ownsCredential: boolean; credential?: CredentialsEntity }> {
-		const sharing = await this.getSharing(userId, credentialId, ['credentials']);
+		const sharing = await this.getSharedCredentials(user, credentialId, ['credentials'], false);
```
Nitpick: A standalone boolean arg is not very readable, ideally in an object?

ivov commented on packages/cli/src/credentials/credentials.service.ts:
```diff
@@ -1,31 +1,91 @@
 		};
 
+		if (!allowGlobalOwner || user.globalRole.name !== 'owner') {
+			// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
```
`FindOneOptions` accepts a generic, so this exception might not be needed?

valya replied:
I've added a comment explaining this but basically on non-EE versions the owner should be able to access and delete credentials they don't own. This is the expected usual behaviour and was an accidental fix for a bug ticket.

ivov commented on packages/cli/src/credentials/credentials.service.ts:
```diff
@@ -1,31 +1,91 @@
+			// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+			options.where.user = { id: user.id };
+		}
+
```
If `relations` is an empty array, the join still happens? If so, we should check for length.

ivov commented on packages/cli/src/credentials/credentials.service.ts:
```diff
@@ -36,4 +96,165 @@ export class CredentialsService {
+		return newCredentialData;
+	}
+
+	static async getEncryptionKey(): Promise<string> {
+		let encryptionKey: string;
+		try {
+			encryptionKey = await UserSettings.getEncryptionKey();
+		} catch (error) {
+			throw new ResponseHelper.ResponseError(
+				RESPONSE_ERROR_MESSAGES.NO_ENCRYPTION_KEY,
+				undefined,
+				500,
+			);
+		}
+		return encryptionKey;
+	}
```
I would just return inside the try block.

ivov commented on packages/cli/src/credentials/credentials.service.ts:
```diff
@@ -36,4 +96,165 @@ export class CredentialsService {
+		data: CredentialRequest.CredentialProperties,
+	): Promise<CredentialsEntity> {
+		// Make a copy so we can delete the provided ID
+		const dataCopy = clone(data);
+		delete dataCopy.id;
```
Or `const { id, ...rest } = data;` and drop the `id`.

valya replied:
I've updated the comment here to explain why I do a deep copy here instead of a spread. It's to avoid side effects from modifying the `nodesAccess`s slightly further down the function.

valya replied:
I've changed this after some changes made from Ben's review. It's now doing assignment spread.

ivov commented on packages/cli/src/credentials/credentials.service.ts:
```diff
@@ -36,4 +96,165 @@ export class CredentialsService {
+	): Promise<CredentialsEntity> {
+		// To avoid side effects
+		const newCredential = new CredentialsEntity();
+		Object.assign(newCredential, credential);
+
+		Object.assign(newCredential, encryptedData);
```
`Object.assign` takes multiple sources so you can combine.

ivov commented on packages/cli/test/integration/credentials.test.ts:
```diff
@@ -59,14 +59,14 @@ test('POST /credentials should create cred', async () => {
 	expect(name).toBe(payload.name);
 	expect(type).toBe(payload.type);
-	expect(nodesAccess[0].nodeType).toBe(payload.nodesAccess[0].nodeType);
+	expect(nodesAccess[0].nodeType).toBe(payload.nodesAccess![0].nodeType);
```
We can `fail()` instead of asserting.

Commit:
:refactor: changes to credentials controller and service from review

 - removed credentials from service function names
 - made relations list optional
 - put allowGlobalOwner in options objects
 - check length of relations array so join doesn't happen if empty
 - update some comments to further explain rationale
 - remove unneeded `Object.assign`
 - remove non-null assertion from test

Commit:
Merge remote-tracking branch 'origin/user-management-p2' into n8n-3959-update-existing-credentials-routes-12

BHesseldieck commented on packages/cli/src/credentials/credentials.service.ts:
```diff
@@ -36,4 +96,162 @@ export class CredentialsService {
+		const dataCopy = clone(data);
+		delete dataCopy.id;
+
+		const newCredentials = new CredentialsEntity();
+		Object.assign(newCredentials, dataCopy);
```
this saves us the merge
```suggestion
		const newCredentials = Db.collections.Credentials.create(dataCopy);
```

BHesseldieck commented on packages/cli/src/credentials/credentials.service.ts:
```diff
@@ -36,4 +96,162 @@ export class CredentialsService {
+		data: CredentialRequest.CredentialProperties,
+		decryptedData: ICredentialDataDecryptedObject,
+	): Promise<CredentialsEntity> {
+		const updateData = new CredentialsEntity();
+		Object.assign(updateData, data);
```
same as previous comment
```suggestion
		const updateData = Db.collections.Credentials.create(data);
```

Commit:
:recycle: move filtered credentials selected fields to variable

valya commented on packages/cli/src/credentials/credentials.controller.ee.ts:
```diff
@@ -28,7 +28,7 @@ EECredentialsController.post('/:id/share', async (req: CredentialRequest.Share,
 	const { shareeId } = req.body;
 
-	const isOwned = EECredentials.isOwned(req.user.id, id);
+	const isOwned = EECredentials.isOwned(req.user, id);
```
I changed it to fit the changes I'd made to `getShared` which requires the whole User. I can change `getShared` to accept either a User or an ID if that'd be better?

BHesseldieck replied:
ah sorry just saw this now, let's take your version 👍🏼

Commit:
:recycle: remove unneeded merges in credentials service

BHesseldieck commented on packages/cli/src/credentials/credentials.service.ts:
```diff
@@ -1,31 +1,91 @@
+		user: User,
 		credentialId: number | string,
-		relations?: string[],
+		relations: string[] | undefined = ['credentials'],
```
but how can I now get a `SharedCredential` without also loading the `credentials`?

valya replied:
Either pass in `undefined` explicitly or an empty array.

BHesseldieck replied:
I think just the empty array would work, otherwise the relations argument defaults to `['credentials']` 🤔

valya replied:
Ahh, right. I was thinking of Python (passing in an explicit `None`).

BHesseldieck committed:
Merge branch 'user-management-p2' into n8n-3959-update-existing-credentials-routes-12
